errors/errors.go

```diff
@@ -11,19 +15,98 @@ import (
 	"github.com/miekg/dns"
 )
 
+var log = clog.NewWithPlugin("errors")
+
+type pattern struct {
+	ptimer  unsafe.Pointer
+	count   uint32
+	period  time.Duration
+	pattern *regexp.Regexp
+}
+
+func (p *pattern) timer() *time.Timer {
+	return (*time.Timer)(atomic.LoadPointer(&p.ptimer))
+}
+
+func (p *pattern) setTimer(t *time.Timer) {
+	atomic.StorePointer(&p.ptimer, unsafe.Pointer(t))
+}
+
 // errorHandler handles DNS errors (and errors from other plugin).
-type errorHandler struct{ Next plugin.Handler }
+type errorHandler struct {
+	patterns []*pattern
+	eLogger  func(int, string, string, string)
+	cLogger  func(uint32, string, time.Duration)
```

Comment:
Nit picking...

It seems unnecessary to call `errorLogger()` and `consLogger()` through these struct fields. IMHO, we can just call them directly from `ServerDNS()`.

Alternatively, if the intention is so that we can enhance those functions to do something more clever, then we should make them methods, i.e.

`
func (h *errorHandler) errorLogger(...)
`

Reply:
This was done for easier testing. We can replace these logger functions with some fake versions as needed for tests.
In the same time, this approach doesn't lead to performance penalties in production code

Reply:
I see.

Another common way of doing this (to support writing tests) is to assign the function to a global (package scope) `var`. The unit test can then assign a test function to the `var`.

Just for your consideration - I don't have a strong opinion.

Reply:
I seek in the code the usage of clog.Errof(...) and it is used by several plugins.
So it is not bad (IMO) that we have "plugin/errors:" in the log from that speific plugin. 
It shows it is a regular error of DNS Processing, not a wrong behavior of the code itself.